earthly/docs/Earthfile

```diff
@@ -0,0 +1,175 @@
+        graphviz
+
+    # Install Poetry to manage python.
+    RUN curl -sSL https://install.python-poetry.org | python3 -
+    RUN poetry config installer.max-workers 10
+
+    # Copy our dependencies.
+    COPY pyproject.toml .
+    COPY poetry.lock .
+    COPY start.sh .
+
+    # Install it all with poetry
+    RUN poetry install --only main --no-root
+
+    # Trust directory, required for git >= 2.35.2
+    RUN git config --global --add safe.directory /docs &&\
+        git config --global --add safe.directory /site
+
+    # Carry the Kroki Docker Compose with this image.
+    WORKDIR /kroki
+    COPY docker-compose.kroki.yml ./docker-compose.yml
+
+    # Set working directory
+    WORKDIR /docs
+
+    # Expose MkDocs development server port
+    EXPOSE 8000
+
+    # Start development server by default
+    ENTRYPOINT ["/poetry/start.sh"]
+    CMD ["serve", "--dev-addr=0.0.0.0:8000"]
+
+    # Finish building the container, and publish to the registry
+    IF [ "$registry" = "" ]
+        ARG registry_final=$registry
+    ELSE
+        ARG registry_final=${registry}/
+    END
+
+    SAVE ARTIFACT /kroki/docker-compose.yml /kroki/docker-compose.yml
+
+    SAVE IMAGE --push ${registry_final}mkdocs-material:${tag}
+
+MKDOCS_BUILD:
+    COMMAND
+    ARG src
+    ARG dest
+
+    # Ensure the mkdocs tools are built and up-to-date.
+    BUILD +mkdocs-material
+
+    FROM earthly/dind:alpine
+    WORKDIR /build
+    COPY +mkdocs-material/kroki/docker-compose.yml .
+
+    # Copy the src
+    COPY --dir $src /docs
+
+    # Docs need Kroki to build diagrams, so make sure its running.
+    WITH DOCKER --compose docker-compose.yml --load mkdocs=+mkdocs-material
+        RUN docker run --rm -v /docs:/docs --network=default_mkdocs mkdocs build
+    END
+
+    SAVE ARTIFACT /docs/site AS LOCAL $dest
+
+MKDOCS_SERVE:
+    COMMAND
+    ARG src
+    ARG port
+
+    # Ensure the mkdocs tools are built and up-to-date.
+    BUILD +mkdocs-material
+
+    FROM earthly/dind:alpine
+    WORKDIR /build
+    COPY +mkdocs-material/kroki/docker-compose.yml .
+
+    # Copy the src
+    COPY --dir $src /docs
+
+    # Docs need Kroki to build diagrams, so make sure its running.
+    WITH DOCKER --compose docker-compose.yml --load mkdocs=+mkdocs-material
+        RUN --privileged docker run --rm -p $port:8000 -v /docs:/docs --network=default_mkdocs mkdocs
+    END
+
+MDLINT:
+    # Linting is done with MarkdownLint CLI2
+    # See: https://github.com/DavidAnson/markdownlint-cli2
+    COMMAND
+
+    ARG fix
+    ARG src
+
+    RUN --no-cache echo SRC: $src
+
+    # cspell: words davidanson
+    WITH DOCKER --pull davidanson/markdownlint-cli2-rules:next
+        #RUN --no-cache docker run --rm -v $src:/workdir --entrypoint sh davidanson/markdownlint-cli2-rules:next -c "ls -al; ls -al docs"
+        RUN --no-cache docker run --rm -v $src:/workdir davidanson/markdownlint-cli2-rules:next "**/*.md" --config .markdownlint-cli2.jsonc $fix
```

Comment:
What is `$fix` doing here? 

This seems really fragile. I'd almost rather vendor the ruleset and then update it as needed. It makes this simpler and also prevents surprises.

Reply:
So mdlint can autofix certain violations.  Especially handy when you don;t have the right file spacing and what not.
So `$fix` was my hacky way of using one rule to do two things.  `lint only`  or `lint and fix`.
On a developers machine, they probably always want to run `lint and fix` in CI we only want `lint`.
I will see if I can think of a cleaner way.

Reply:
I do need the $fix option, but I made it so it can only accept the one allowed value (or blank).  Anything else will fail the build.